File: exp-14/q1-kruskals/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type Edge struct {
	u, v, w int
}

func find(parent []int, x int) int {
	if parent[x] != x {
		parent[x] = find(parent, parent[x])
	}
	return parent[x]
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	next := func() int {
		scanner.Scan()
		var x int
		fmt.Sscanf(scanner.Text(), "%d", &x)
		return x
	}

	n := next()
	m := next()

	edges := make([]Edge, m)
	for i := 0; i < m; i++ {
		edges[i].u = next()
		edges[i].v = next()
		edges[i].w = next()
	}

	sort.Slice(edges, func(i, j int) bool {
		return edges[i].w < edges[j].w
	})

	parent := make([]int, n+1)
	for i := 1; i <= n; i++ {
		parent[i] = i
	}

	fmt.Println("Edges in MST:")
	total, count := 0, 0
	for _, e := range edges {
		if count == n-1 {
			break
		}
		ru := find(parent, e.u)
		rv := find(parent, e.v)
		if ru != rv {
			parent[ru] = rv
			fmt.Printf("%d - %d (%d)\n", e.u, e.v, e.w)
			total += e.w
			count++
		}
	}

	fmt.Printf("Total weight of MST: %d\n", total)
}

File: exp-14/q1-kruskals/Solution.java
import java.util.*;

public class Solution {

    static int[] parent;

    static int find(int x) {
        if (parent[x] != x)
            parent[x] = find(parent[x]);
        return parent[x];
    }

    static void unionSet(int a, int b) {
        parent[a] = b;
    }

    static class Edge {
        int u, v, w;
        Edge(int u, int v, int w) {
            this.u = u;
            this.v = v;
            this.w = w;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        List<Edge> edges = new ArrayList<>();
        for (int i = 0; i < m; i++)
            edges.add(new Edge(sc.nextInt(), sc.nextInt(), sc.nextInt()));

        edges.sort(Comparator.comparingInt(e -> e.w));

        parent = new int[n + 1];
        for (int i = 1; i <= n; i++)
            parent[i] = i;

        System.out.println("Edges in MST:");
        int total = 0, count = 0;
        for (Edge e : edges) {
            if (count == n - 1) break;
            int ru = find(e.u);
            int rv = find(e.v);
            if (ru != rv) {
                unionSet(ru, rv);
                System.out.println(e.u + " - " + e.v + " (" + e.w + ")");
                total += e.w;
                count++;
            }
        }

        System.out.println("Total weight of MST: " + total);
        sc.close();
    }
}
